CLOUDFLAREAPI: Initial support for CF_WORKER_ROUTE (without tests).

fdcastel opened the pull request:
Initial implementation of CF_WORKER_ROUTE.

Adds [Cloudflare Worker routes](https://developers.cloudflare.com/workers/platform/routes) support via new `CF_WORKER_ROUTE` function.

No tests, as I have yet to understand the tests infrastructure. But the code is working as I did some manual tests with a personal CF account (Create, Modify and Delete routes are working as expected).

Thanks to @tlimoncelli for guidance and incentive.

fdcastel commented:
Added initial tests, similar to `CF_REDIRECT` ones.

Guidance required for the integration tests (I didn't find where the expected result should be).

tlimoncelli commented on docs/_functions/domain/CF_WORKER_ROUTE.md:
```diff
@@ -0,0 +1,29 @@
+  - script
+---
+
+`CF_WORKER_ROUTE` uses [Cloudflare Workers](https://developers.cloudflare.com/workers/) 
```
s/uses/uses the/

tlimoncelli commented on docs/_functions/domain/CF_WORKER_ROUTE.md:
```diff
@@ -0,0 +1,29 @@
+for a given domain.
+
+If _any_ `CF_WORKER_ROUTE` function is used then `dnscontrol` will manage _all_ 
+Worker Routes for the domain.
```
If this means it will delete routes that were added via the web portal, be extremely clear:

```
for the domain. To be clear: this means it will delete existing routes that were created outside of DNSControl.
```

tlimoncelli commented on docs/_functions/domain/CF_WORKER_ROUTE.md:
```diff
@@ -0,0 +1,29 @@
+
+WARNING: This interface is not extensively tested. Take precautions such as making
+backups and manually verifying `dnscontrol preview` output before running
+`dnscontrol push`. This is especially true when mixing Worker Routes that are
```
Can we mix worker routes that aren't managed by DNSControl?  If not, this last sentence should make that clear.

fdcastel replied:
To be frank, this was copied from CF_REDIRECT. The same logic applies here.

I confess I don't like the original text too I don't think anybody would use DNS Control and manually set some routes (since any route manually included will be removed in next push). Or I am missing something?

fdcastel replied:
I think I got it. The original text was written that way because there are OTHER kind of page rules not touched by DNSControl (only "Forwarding URL" rules are managed by it).

I'm removing this sentence.

tlimoncelli commented on models/record.go:
```diff
@@ -34,12 +34,14 @@ import (
 //     NO_PURGE
 //     NS1_URLFWD
 //     PAGE_RULE
+//     WORKER_ROUTE
```
Is there a "WORKER_ROUTE"?

fdcastel replied:
Yes. `WORKER_ROUTE` is for `CF_WORKER_ROUTE` what a `PAGE_RULE` is for a `CF_REDIRECT`/`CF_TEMP_REDIRECT`.

tlimoncelli commented on providers/cloudflare/cloudflareProvider.go:
```diff
@@ -227,6 +234,11 @@ func (c *cloudflareProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*m
 				Msg: d.String(),
 				F:   func() error { return c.deletePageRule(ex.Original.(*pageRule).ID, id) },
 			})
+		} else if ex.Type == "WORKER_ROUTE" {
```
Let's consistently use `CF_WORKER_ROUTE` and not `WORKER_ROUTE` (or are they 2 different things?)

fdcastel replied:
2 different things. See last comment.

tlimoncelli commented on providers/cloudflare/cloudflareProvider.go:
```diff
@@ -245,6 +257,11 @@ func (c *cloudflareProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*m
 				Msg: d.String(),
 				F:   func() error { return c.createPageRule(id, des.GetTargetField()) },
 			})
+		} else if des.Type == "WORKER_ROUTE" {
```
CF_WORKER_ROUTE?

tlimoncelli commented on providers/cloudflare/cloudflareProvider.go:
```diff
@@ -265,6 +282,11 @@ func (c *cloudflareProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*m
 				Msg: d.String(),
 				F:   func() error { return c.updatePageRule(ex.Original.(*pageRule).ID, id, rec.GetTargetField()) },
 			})
+		} else if rec.Type == "WORKER_ROUTE" {
```
CF_WORKER_ROUTE?

tlimoncelli commented on providers/cloudflare/cloudflareProvider.go:
```diff
@@ -423,6 +445,16 @@ func (c *cloudflareProvider) preprocessConfig(dc *models.DomainConfig) error {
+				return fmt.Errorf("invalid data specified for cloudflare worker record")
+			}
+			rec.TTL = 1
+			rec.Type = "WORKER_ROUTE"
```
CF_WORKER_ROUTE

tlimoncelli commented on providers/cloudflare/rest.go:
```diff
@@ -505,6 +507,80 @@ func (c *cloudflareProvider) sendPageRule(endpoint, method string, data string)
+
+func (c *cloudflareProvider) sendWorkerRoute(endpoint, method string, data string) error {
+	// $PATTERN,$SCRIPT
+	parts := strings.Split(data, ",")
```
return an error if `len(parts) != 2`

tlimoncelli commented on docs/_providers/cloudflare.md:
```diff
@@ -170,3 +170,20 @@ Notice a few details:
 4. if _any_ `CF_REDIRECT` or `CF_TEMP_REDIRECT` functions are used then `dnscontrol` will manage _all_ "Forwarding URL" type Page Rules for the domain. Page Rule types other than "Forwarding URL” will be left alone.
+
+## Worker routes
+The Cloudflare provider can manage Worker Routes for your domains. Simply use the `CF_WORKER_ROUTE` function passing the route pattern and the worker name:
```
Repeat the same paragraph as before:

```
If _any_ `CF_WORKER_ROUTE` function is used then `dnscontrol` will manage _all_ 
Worker Routes for the domain. To be clear: this means it will delete existing routes that
were created outside of DNSControl.
```

tlimoncelli commented on providers/cloudflare/rest.go:
```diff
@@ -505,6 +507,83 @@ func (c *cloudflareProvider) sendPageRule(endpoint, method string, data string)
+	script := parts[1]
+	wr := &workerRoute{
+		Pattern: pattern,
+		Script:  script}
```
Please format the last item like this so that diff's in the future will be smaller.

```
    Script: script,
}
```

tlimoncelli reviewed (commented):
Just a few more comments.

* Please rebase
* Please add a test to integrationTest/integration_test.go

fdcastel commented:
About the integration tests... I am somewhat lost. 😢 

How do we test this? How could I check if the test is passing or not (won't I need some API credentials to run it locally?). 

I'm surely missing something here 😅

tlimoncelli commented:
Here is how to add a simple integration test: https://github.com/StackExchange/dnscontrol/pull/1250

To run the integration tests you'll need a test domain:

```
export CLOUDFLAREAPI_DOMAIN='dnscontroltest-cf.com'      <<< CHANGE TO YOUR DOMAIN
export CLOUDFLAREAPI_TOKEN='REDACTED'
go test -v -verbose -provider CLOUDFLAREAPI
```

Even without that PR, I've found a small bug.  It looks like the code assumes that there will be a worker route list.

```
tlimwork5:~/git/dnscontrol/integrationTest$ go test -v -verbose -provider CLOUDFLAREAPI
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/dnscontroltest-cf.com
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Clean_Slate:Empty
{"success":false,"errors":[{"code":10000,"message":"Authentication error"}]}

    integration_test.go:193: runTests: failed fetching worker route list from cloudflare: bad status code from cloudflare: 403 not 200
=== RUN   TestDNSProviders/dnscontroltest-cf.com/00:GeneralACD:Create_an_A_record
{"success":false,"errors":[{"code":10000,"message":"Authentication error"}]}

    integration_test.go:193: runTests: failed fetching worker route list from cloudflare: bad status code from cloudflare: 403 not 200
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Post_cleanup:Empty
{"success":false,"errors":[{"code":10000,"message":"Authentication error"}]}
```

tlimoncelli commented:
FYI: The automated tests (https://github.com/StackExchange/dnscontrol/pull/1243/checks) are skipping CloudFlare because it can't find any credentials.  Github Actions uses the credentials from the fdcastel/dnscontrol repo and there are none.  Once it is merged to StackExchange/dnscontrol it will have access to the credentials in StackExchange/dnscontrol and the tests will run. (this is to protect someone from writing a PR that steals the creds.)

fdcastel commented:
I'm still missing something. Even after pulling your sample code and settings my env vars like:

```powershell
$env:CLOUDFLAREAPI_DOMAIN='my-cf-domain.com'
$env:CLOUDFLAREAPI_KEY='my-api-key'
$env:CLOUDFLAREAPI_USER='my-api-user'
```
A call to 
```powershell
go test -v -verbose -provider CLOUDFLAREAPI
```

gives me

```
?   	github.com/StackExchange/dnscontrol/v3	[no test files]
``` 

_Running on Windows 10 / Powershell_

fdcastel commented:
Just to be clear: I already read https://stackexchange.github.io/dnscontrol/byo-secrets and it seems clear to me. 

But I'm trying to run the tests locally on my machine before I put them into the CI. 

Is this possible or I misunderstood something? (i.e., the tests would only run inside Github CI engine?)

fdcastel commented:
Got it!

```powershell
go test -v ./integrationTest -verbose -provider CLOUDFLAREAPI  
```

Tests are running now 🙌 . Working on it.

fdcastel commented:
There is some way -- in the current test infrastructure -- to run additional code _before_ some `testgroup` starts to run?

Rationale: I have the following test:

```javascript
  testgroup("CF_WORKER_ROUTE",
    only("CLOUDFLAREAPI"),
    tc("simple", cfWorkerRoute("cnn.**current-domain-no-trailing**/*", "cnn-worker")),
    tc("changeScript", cfWorkerRoute("cnn.**current-domain-no-trailing**/*", "msnbc-worker")),
```

But before it, I need to call an additional Cloudflare api to _create_ `cnn-worker` and `msnbc-worker` workers. 

Without this, the API calls under test will fail with a

```json
[{"code":10019,"message":"workers.api.error.invalid_route_script_missing"}]
```

fdcastel commented:
✔️ Rebase done. 
✔️ Integration tests done.
✔️ Requested changes done.
❗ Integration tests *need two cloudflare workers* (named `cnn-worker` and `msnbc-worker`) on cloudflare account to pass.

tlimoncelli commented:
> There is some way -- in the current test infrastructure -- to run additional code _before_ some `testgroup` starts to run?
> 
> Rationale: I have the following test:
> 
> ```js
>   testgroup("CF_WORKER_ROUTE",
>     only("CLOUDFLAREAPI"),
>     tc("simple", cfWorkerRoute("cnn.**current-domain-no-trailing**/*", "cnn-worker")),
>     tc("changeScript", cfWorkerRoute("cnn.**current-domain-no-trailing**/*", "msnbc-worker")),
> ```
> 
> But before it, I need to call an additional Cloudflare api to _create_ `cnn-worker` and `msnbc-worker` workers.
> 
> Without this, the API calls under test will fail with a
> 
> ```json
> [{"code":10019,"message":"workers.api.error.invalid_route_script_missing"}]
> ```

There isn't a uniform way to do setup code.  However....

Around line 55 of integration_test.go we do some "just for CloudFlare" actions.  Maybe that could call a function that creates the workers if they don't already exist?  Please create them with names that won't overlap real workers... maybe dnsctrl_integrationtest_foo and dnsctrl_integrationtest_bar ?

Tom

fdcastel commented:
Minor update: Yet working on it. I should update this PR in a few days. Thanks, @tlimoncelli !

fdcastel commented:
_Sorry for the long delay. It has been a long week around here._

It's almost done. However, **I'm having a hard time making Cloudflare-specific API calls from `integration_test.go`.** It seems wrong to call provider-specific code here. I don't event have the context for Cloudflare account, creds, etc.

One solution would be to add a `PrepareTests()` in `DNSProvider` interface. However this would add a lot of empty boilerplate code for every other provider which doesn't need this. Not good, IMHO.

Could we use `preprocess_test.go` for the test workers creation? There is *any* guarantee that this code would run *before* the tests from `integration_test.go`?

fdcastel commented:
Pushed what I got so far. Looking for advice.

I added `TestPreprocess_CreateTestWorkers` in `preprocess_test.go` but it is not being called before the integration tests.

I also added two new env vars `CLOUDFLAREAPI_ACCOUNTID` and `CLOUDFLAREAPI_ACCOUNTNAME` which were already present in `creds.json` but without any way to set them via environment. 

AccountName is not used, actually. However, there is [a test in `cloudflareProvider.go`](https://github.com/StackExchange/dnscontrol/blob/master/providers/cloudflare/cloudflareProvider.go#L466) which enforces both vars to be set.

tlimoncelli commented:
Hi there!

Sorry for the delay.  It's been a crazy week!

A few things:

1.  I haven't been able to get this running since I can't figure out the permissions.  Do you know what perms to give this API key to enable it to use workers?

2.  I've noticed that my API key doesn't have access to Workers at all (maybe because its a free account?). I tried to update the API's permissions but I get:
```
    integration_test.go:193: runTests: failed fetching worker route list from cloudflare: bad status code from cloudflare: 403 not 200
```

The code either needs to automagically detect if Workers is disabled and be silent or there needs to be a feature-flag to enable Workers.  Otherwise this will be a breaking change... people that don't use workers can't break and can't be asked to modify permissions on the API key just because of a new feature.

I don't know how to detect if workers is enabled, but I'm making a PR that adds the feature flag in case you go that route: https://github.com/fdcastel/dnscontrol/pull/1/files

3. `preprocess_test.go` is a unit test and isn't expected to run as part of the integration tests.  Instead, it runs when you do "go test -count=1 ./..." from `providers/cloudflare`.

The right place for `TestPreprocess_CreateTestWorkers` is in `integrationTest/`.  Yes, that's a layering violation but... that's life.  I would move it to `integrationTest/cloudflare.go` and call it cloudflareSetupWorkers().  You can call it from  `integration_test.go` around line 53 where we do some cloudflare-specific things already.

At least I think that will work. I'm not entirely sure as I can't test it myself until I fix my own permission problem.

I'm in the US/Eastern timezone if you would like to screenshare and work on this in real time. 

Tom

fdcastel commented:
Thanks, Tom, for the insights and the pragmatism :wink: - I will return to this in a few hours.

About your points:

1. I'm using Cloudflare **API Keys** (which, as I understand, gives me _sudo-like_ access to the account) instead of **API Tokens** (which allows you more granular control about what can be accessed). That said, after a quick glimpse at the options available in Cloudflare dashboard, I believe these are what you will need:

![image](https://user-images.githubusercontent.com/629408/133000156-887886a5-cbd1-4e81-a434-d4ec549f136b.png)

2. You definitely can use Workers on a Free account (my own test domain is on this very same plan). So, the problem is probably permission-related. I will investigate a bit more later to know what specific permissions are needed for token-related access (but I would guess that the above image is very close).
 
3. Perfect! Got it!

fdcastel commented:
Rebase to latest release.

fdcastel commented:
Added feature flag for CF Workers per @tlimoncelli suggestion.

Porting tests back to `integration_tests/` now -- https://www.youtube.com/watch?v=TnlPtaPxXfc

fdcastel commented:
All done! Integration tests are creating test workers now.

Two considerations:

- I added 'testing' module in `cloudflare` module. Surely I'm not proud of this and would welcome any suggestions to improve (if deemed necessary).

- Tom: I believe we now have a better understanding about [that comment](https://github.com/StackExchange/dnscontrol/blob/master/integrationTest/integration_test.go#L49) of `integration_test.go` :wink: (needed to avoid breaking changes?). We could improve it explaining the rationale behind the feature flags.

fdcastel commented:
I ran all tests twice (first without test workers and then with them) in a test domain and everything looks good to me.

Save for further coding improvements (see last comment) I would consider the work now done.

Also, if agreed, I would like to rewrite the commit history to put the commits in a better shape before merge.

tlimoncelli commented:
Looks very good so far!

I do want to find a way to make legacy domains "just work".  Here's the one I use for tests:

```
$ go test -v -verbose -provider CLOUDFLAREAPI
=== RUN   TestDNSProviders
    cloudflareProvider.go:699: [{"code":7003,"message":"Could not route to /accounts/workers/scripts/dnscontrol_integrationtest_cnn, perhaps your object identifier is invalid?"},{"code":7000,"message":"No route for that URI"}]
--- FAIL: TestDNSProviders (0.92s)
=== RUN   TestDualProviders
    cloudflareProvider.go:699: [{"code":7003,"message":"Could not route to /accounts/workers/scripts/dnscontrol_integrationtest_cnn, perhaps your object identifier is invalid?"},{"code":7000,"message":"No route for that URI"}]
--- FAIL: TestDualProviders (0.87s)
FAIL
exit status 1
FAIL	github.com/StackExchange/dnscontrol/v3/integrationTest	2.041s

```

fdcastel commented:
Sorry, Tom. I should have stressed this :

**To make the tests work you must now set `CLOUDFLAREAPI_ACCOUNTID` and `CLOUDFLAREAPI_ACCOUNTNAME`.** These are required to the API calls which create the workers.

That `/accounts/workers/scripts/dnscontrol_integrationtest_cnn` path on your error message should be `/accounts/<accoundid>/workers/scripts/dnscontrol_integrationtest_cnn`.

Maybe I should add a test and emit a `Fail()` in this case (missing account id and account name) to warn the user?

P.S: AccountName is not used at all. It is only needed because of another check. Please see [this comment](https://github.com/StackExchange/dnscontrol/pull/1243#issuecomment-917439566).

tlimoncelli commented:
Hi there!

I'm sorry but I'm having trouble getting this to run on a legacy domain.

It is very important that we make this change in an upwards compatible manner.  Currently we get errors like:

```
$ go test -v -verbose -provider CLOUDFLAREAPI
=== RUN   TestDNSProviders
    cloudflareProvider.go:699: [{"code":7003,"message":"Could not route to /accounts/workers/scripts/dnscontrol_integrationtest_cnn, perhaps your object identifier is invalid?"},{"code":7000,"message":"No route for that URI"}]
--- FAIL: TestDNSProviders (0.96s)
=== RUN   TestDualProviders
    cloudflareProvider.go:699: [{"code":7003,"message":"Could not route to /accounts/workers/scripts/dnscontrol_integrationtest_cnn, perhaps your object identifier is invalid?"},{"code":7000,"message":"No route for that URI"}]
--- FAIL: TestDualProviders (0.80s)
FAIL
exit status 1
FAIL	github.com/StackExchange/dnscontrol/v3/integrationTest	2.087s
```

How about we add a feature-flag?  For example, with "CLOUDFLARE_WORKERS=1" the new code runs. If it isn't set, the new code silently disables itself.

Tom

fdcastel commented:
Sorry, Tom. I don't understand. Did you see my [last message](https://github.com/StackExchange/dnscontrol/pull/1243#issuecomment-923488723)? 

According to your error message:

```
 [{"code":7003,"message":"Could not route to /accounts/workers/scripts/dnscontrol_integrationtest_cnn, perhaps your object identifier is invalid?"},{"code":7000,"message":"No route for that URI"}]
```

It seems that the `CLOUDFLAREAPI_ACCOUNTID` wasn't set during your test. I can only presume this since the path above should be `/accounts/<ACCOUNT-ID>/workers/...`.

To make the things clearer, this is how I ran the tests:

```powershell
go build 

$env:CLOUDFLAREAPI_DOMAIN='my-test-domain.com'
$env:CLOUDFLAREAPI_KEY='my-api-key'
$env:CLOUDFLAREAPI_USER='my@email.com'

$env:CLOUDFLAREAPI_ACCOUNTID='my-account-id'
$env:CLOUDFLAREAPI_ACCOUNTNAME='MyAccountName'

go test -v ./integrationTest -verbose -provider CLOUDFLAREAPI  
```

The `AccountId` can be seen on the browser url after the first logon to https://dash.cloudflare.com. You could also see it on "Workers" section from Cloudflare administration panel:

![image](https://user-images.githubusercontent.com/629408/134259795-31bbbe36-730a-4474-a808-f66ac9711ac4.png)

Please tell me how this has worked for you. I don't see the need for an _additional_ feature flag given that we already added one (`manage_workers`) exactly in the name of backward compatibility.

tresni commented:
#1267 and this will conflict. Happy to help build this on top of #1267 if that makes sense.

fdcastel commented:
Whoa! Fantastic work, @tresni ! 😄 

Working on it right now.

fdcastel commented:
All done! Lots of code slashed 😄. Commit history rewrite. All tests passing.

@tresni could you kindly review https://github.com/StackExchange/dnscontrol/pull/1243/commits/b55278140f63e0f95a3418bcd61b810c731ae08f ? I need `cfClient.AccountID` to be defined and I believe that's the right place (but please correct me if I'm wrong).

@tlimoncelli Did you manage to make [the tests pass](https://github.com/StackExchange/dnscontrol/pull/1243#issuecomment-924072372) with your domain? Do we need anything else to merge this? (Please? Pretty please? 😉)

tlimoncelli commented:
Sadly I'm still getting an error:

```
$ go test -v -verbose -provider CLOUDFLAREAPI
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/dnscontroltest-cf.com
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Clean_Slate:Empty
    integration_test.go:197: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/00:GeneralACD:Create_an_A_record
    integration_test.go:197: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Post_cleanup:Empty
    integration_test.go:197: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/01:WildcardACD:Create_wildcard
    integration_test.go:197: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Post_cleanup:Empty#01
    integration_test.go:197: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
```

tresni commented:
@fdcastel thats probably a better way to handle it then I did later (in the one place we used it previously.) I’ll check this locally and see if I can help you clean up the tests/get them working.

tlimoncelli commented:
We're getting closer!

Getting this to work in a way that is upwards compatible with existing users is... difficult.  I'm experimenting with some ideas.  The first is to simply have a creds.json variable that will be a feature-flag, enabling all this stuff.  It's kind of a big hammer for a small problem, but it works.

My work in progress is https://github.com/StackExchange/dnscontrol/pull/1269

tresni commented on providers/cloudflare/cloudflareProvider.go:
```diff
@@ -638,3 +680,19 @@ func (c *cloudflareProvider) EnsureDomainExists(domain string) error {
 	return err
 }
+
+// PrepareCloudflareWorkers creates Cloudflare Workers required for CF_WORKER_ROUTE tests.
```
I feel like this should move to the integrationTests source to prevent bringing `testing` into the production build.

tresni commented on providers/cloudflare/rest.go:
If we moved this into a `something_test.go` file would the build still work.  It's not something we are actually using in production code, but I totally understand why it's in the cloudflare provider space.

tresni commented:
@tlimoncelli this PR seems to work fine for me.  I don't use cloudflare workers normally and everything runs as expected (I don't set accountId and `"manage_workers"` defaults to `false`.)  This is fully backwards compatible with existing code/deployments as is.

@fdcastel  Here are the changes I made to work with latest master and get `testing` out of the main code. Still don't like that there is test setup code in here, but I tried everything I could think of, no easy way to move it out completely unless we make `cloudflareProvider` public (i.e. `CloudflareProvider`) as that would allow us to move all the setup code into intergration_test.go.

Also, this doesn't cleanup the workers after-the-fact which may not be great, but I don't think it's really hurting anything.


```patch
diff --git a/integrationTest/integration_test.go b/integrationTest/integration_test.go
index a8250c51..28e749e4 100644
--- a/integrationTest/integration_test.go
+++ b/integrationTest/integration_test.go
@@ -70,8 +70,12 @@ func getProvider(t *testing.T) (providers.DNSServiceProvider, string, map[int]bo
 			}
 		}
 
-		// Cloudflare only. Will do nothing if provider != *cloudflareProvider.
-		cloudflare.PrepareCloudflareTestWorkers(t, provider)
+		if name == "CLOUDFLAREAPI" {
+			// Cloudflare only. Will do nothing if provider != *cloudflareProvider.
+			if err := cloudflare.PrepareCloudflareTestWorkers(provider); err != nil {
+				t.Fatal(err)
+			}
+		}
 
 		return provider, cfg["domain"], fails, cfg
 	}
diff --git a/providers/cloudflare/cloudflareProvider.go b/providers/cloudflare/cloudflareProvider.go
index 9ceb4418..e1e4851e 100644
--- a/providers/cloudflare/cloudflareProvider.go
+++ b/providers/cloudflare/cloudflareProvider.go
@@ -6,7 +6,6 @@ import (
 	"log"
 	"net"
 	"strings"
-	"testing"
 
 	"github.com/cloudflare/cloudflare-go"
 	"github.com/miekg/dns/dnsutil"
@@ -506,8 +505,6 @@ func newCloudflare(m map[string]string, metadata json.RawMessage) (providers.DNS
 		api.cfClient.AccountID = m["accountid"]
 	}
 
-	api.cfClient.AccountID = api.AccountID
-
 	if len(metadata) > 0 {
 		parsedMeta := &struct {
 			IPConversions   string   `json:"ip_conversions"`
@@ -674,17 +671,18 @@ func (c *cloudflareProvider) EnsureDomainExists(domain string) error {
 }
 
 // PrepareCloudflareWorkers creates Cloudflare Workers required for CF_WORKER_ROUTE tests.
-func PrepareCloudflareTestWorkers(t *testing.T, prv providers.DNSServiceProvider) {
+func PrepareCloudflareTestWorkers(prv providers.DNSServiceProvider) error {
 	cf, ok := prv.(*cloudflareProvider)
 	if ok {
 		err := cf.createTestWorker("dnscontrol_integrationtest_cnn")
 		if err != nil {
-			t.Fatal(err)
+			return err
 		}
 
 		err = cf.createTestWorker("dnscontrol_integrationtest_msnbc")
 		if err != nil {
-			t.Fatal(err)
+			return err
 		}
 	}
+	return nil
 }
```

tresni commented:
@tlimoncelli to be clear, I ran the integration tests as follows.  I've sanitized the domain I use as it's not really a testing domain :-P

```shell
$ cd integrationTest
$ export CLOUDFLAREAPI_ACCOUNTID=[ACCOUNT_ID]
$ export CLOUDFLAREAPI_DOMAIN=[DOMAIN]
$ export CLOUDFLAREAPI_TOKEN=[TOKEN]
$ go test -v -verbose -provider CLOUDFLAREAPI --start 41
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/Clean_Slate:Empty
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:simple
    integration_test.go:208: CREATE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} cnn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_cnn ttl=1
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:changeScript
    integration_test.go:208: MODIFY WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN}: (cnn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_cnn ttl=1) -> (cnn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1)
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:changePattern
    integration_test.go:208: MODIFY WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN}: (cnn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1) -> (cable.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1)
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:Empty
    integration_test.go:208: DELETE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} cable.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:createMultiple
    integration_test.go:208: CREATE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} cnn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_cnn ttl=1
    integration_test.go:208: CREATE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} msnbc.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:addOne
    integration_test.go:208: CREATE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} api.${CLOUDFLAREAPI_DOMAIN}/cnn/*,dnscontrol_integrationtest_cnn ttl=1
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:changeOne
    integration_test.go:208: MODIFY WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN}: (msnbc.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1) -> (msn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1)
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:deleteOne
    integration_test.go:208: DELETE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} cnn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_cnn ttl=1
=== RUN   TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/Post_cleanup:Empty
    integration_test.go:208: DELETE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} api.${CLOUDFLAREAPI_DOMAIN}/cnn/*,dnscontrol_integrationtest_cnn ttl=1
    integration_test.go:208: DELETE WORKER_ROUTE ${CLOUDFLAREAPI_DOMAIN} msn.${CLOUDFLAREAPI_DOMAIN}/*,dnscontrol_integrationtest_msnbc ttl=1
--- PASS: TestDNSProviders (31.58s)
    --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN} (24.06s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/Clean_Slate:Empty (1.49s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:simple (2.36s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:changeScript (2.69s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:changePattern (2.78s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:Empty (1.06s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:createMultiple (3.17s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:addOne (2.31s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:changeOne (3.45s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/41:CF_WORKER_ROUTE:deleteOne (2.65s)
        --- PASS: TestDNSProviders/${CLOUDFLAREAPI_DOMAIN}/Post_cleanup:Empty (1.99s)
=== RUN   TestDualProviders
    integration_test.go:313: Clearing everything
    integration_test.go:319: Adding nameservers from another provider
WARNING: cloudflare does not support modifying NS records on base domain. ns1.example.com. will not be added.
WARNING: cloudflare does not support modifying NS records on base domain. ns2.example.com. will not be added.
    integration_test.go:322: Running again to ensure stability
WARNING: cloudflare does not support modifying NS records on base domain. ns1.example.com. will not be added.
WARNING: cloudflare does not support modifying NS records on base domain. ns2.example.com. will not be added.
--- PASS: TestDualProviders (9.31s)
PASS
ok  	github.com/StackExchange/dnscontrol/v3/integrationTest	41.469s
$
```

tlimoncelli commented:
I'm getting a build error:

```
$ go install
# github.com/StackExchange/dnscontrol/v3/providers/cloudflare
providers/cloudflare/cloudflareProvider.go:509:30: api.AccountID undefined (type *cloudflareProvider has no field or method AccountID)
```

tresni commented:
@tlimoncelli Yeah, need the patch I included above to make it work with latest master.

fdcastel commented:
Sorry for the delay. Working on it now!

fdcastel commented:
Rebase done with latest master. @tresni patch applied. All tests passing. 🎉

fdcastel commented:
@tresni 

> Still don't like that there is test setup code in here, but I tried everything I could think of, no easy way to move it out completely unless we make `cloudflareProvider` public (i.e. `CloudflareProvider`) as that would allow us to move all the setup code into intergration_test.go.

Fully agreed. I don't like this either. We just took the pragmatic approach, here. 

> Also, this doesn't cleanup the workers after-the-fact which may not be great, but I don't think it's really hurting anything.

Yep. Not ideal. But, also agreed, not a big problem.

Thanks for the patch for the latest master. 👍

fdcastel commented:
@tlimoncelli please let me know if you have any problem with the integration tests.

tresni commented on .github/workflows/build.yml:
```diff
@@ -94,6 +94,8 @@ jobs:
       CLOUDFLAREAPI_TOKEN: ${{ secrets.CLOUDFLAREAPI_TOKEN }}
       CLOUDFLAREAPI_USER: ${{ secrets.CLOUDFLAREAPI_USER }}
+      CLOUDFLAREAPI_ACCOUNTID: ${{ secrets.CLOUDFLAREAPI_ACCOUNTID }}
+      CLOUDFLAREAPI_ACCOUNTNAME: ${{ secrets.CLOUDFLAREAPI_ACCOUNTNAME }}
```
You can remove this, it's not needed anymore :)

tlimoncelli commented:
Here's what I get:

```
tlimwork5:~/git/dnscontrol/integrationTest$ export CLOUDFLAREAPI_DOMAIN='dnscontroltest-cf.com'
tlimwork5:~/git/dnscontrol/integrationTest$ export CLOUDFLAREAPI_TOKEN='redacted'
tlimwork5:~/git/dnscontrol/integrationTest$ echo go test -v -verbose -provider CLOUDFLAREAPI
tlimwork5:~/git/dnscontrol/integrationTest$ go test -v -verbose -provider CLOUDFLAREAPI
=== RUN   TestDNSProviders
    integration_test.go:76: account ID required
--- FAIL: TestDNSProviders (0.00s)
=== RUN   TestDualProviders
    integration_test.go:76: account ID required
--- FAIL: TestDualProviders (0.00s)
FAIL
exit status 1
FAIL	github.com/StackExchange/dnscontrol/v3/integrationTest	0.250s

Since that failed, I tried setting the account id:

tlimwork5:~/git/dnscontrol/integrationTest$ export CLOUDFLAREAPI_ACCOUNTID=a5d8f5aa76a5b5518741f059b5189f9d
tlimwork5:~/git/dnscontrol/integrationTest$ go test -v -verbose -provider CLOUDFLAREAPI
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/dnscontroltest-cf.com
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Clean_Slate:Empty
    integration_test.go:201: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/00:GeneralACD:Create_an_A_record
    integration_test.go:201: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/Post_cleanup:Empty
    integration_test.go:201: runTests: failed fetching worker route list cloudflare: HTTP status 403: Authentication error (10000)
=== RUN   TestDNSProviders/dnscontroltest-cf.com/01:WildcardACD:Create_wildcard
```

However if I run "dnscontrol preview" on Stackoverflow's dnsconfig.js or my personal dnsconfig.js it works just fine.  So, that tells me that it is only the integration tests that are still poking at workers when they shouldn't.

fdcastel commented:
All done! Clean merge. Thanks @tlimoncelli !! 🙌
